Kernel/Prekernel/init.cpp

```diff
@@ -68,11 +68,15 @@ extern "C" [[noreturn]] void init()
     multiboot_module_entry_t* kernel_module = (multiboot_module_entry_t*)(FlatPtr)multiboot_info_ptr->mods_addr;
 
     u8* kernel_image = (u8*)(FlatPtr)kernel_module->start;
-    ElfW(Ehdr)* kernel_elf_header = (ElfW(Ehdr)*)kernel_image;
-    ElfW(Phdr)* kernel_program_headers = (ElfW(Phdr*))((char*)kernel_elf_header + kernel_elf_header->e_phoff);
+    // copy the ELF header and program headers because we might end up overwriting them
+    ElfW(Ehdr) kernel_elf_header = *(ElfW(Ehdr)*)kernel_image;
+    ElfW(Phdr) kernel_program_headers[16];
+    if (kernel_elf_header.e_phnum > sizeof(kernel_program_headers) / sizeof(kernel_program_headers[0]))
```

Comment:
use `array_size()` from AK? unless that's not available here...in which case, maybe just `constexpr size_t program_header_count = 16;` above L73?

Reply:
Fixed. :)